### PR TITLE
Quantization support for CausalVisualLMs

### PR DESCRIPTION
**Pull request by nikita-savelyevv**
# What does this PR do?

Add support for (data-aware) compression of CausalVisualLMs:
- LLaVa (e.g., `llava-hf/llava-v1.6-mistral-7b-hf`)
- NanoLLaVa (e.g, `qnguyen3/nanoLLaVA`)
- MiniCPMV (e.g., `openbmb/MiniCPM-V-2_6`)

When `quantization_config` is given, language model will be compressed according to it. Other model parts, including vision and text embeddings models are compressed to int8_asym.

Example:

optimum-cli
```shell
optimum-cli export openvino --task image-text-to-text -m llava-hf/llava-v1.6-mistral-7b-hf \
--weight-format int4 --dataset contextual --awq --num-samples 32 ./llava_int4_awq
```

Python API
```python
from pathlib import Path

import requests
from PIL import Image
from transformers import AutoConfig, AutoProcessor, AutoTokenizer, TextStreamer

from optimum.intel.openvino import OVModelForVisualCausalLM, OVWeightQuantizationConfig


models_parent_dir = Path("./models")
quantization_config = OVWeightQuantizationConfig(
    bits=4,
    quant_method=OVQuantizationMethod.AWQ,
    dataset="contextual",
    num_samples=32,
    trust_remote_code=False,
    processor=None,
    tokenizer=None,
)


compress = bool(1)
model_arch = "llava"
# model_arch = "nanollava"
# model_arch = "minicpmv"

trc = False
if model_arch == "llava":
    model_id = "llava-hf/llava-v1.6-mistral-7b-hf"
    model_path = models_parent_dir / "llava-v1.6-mistral-7b-hf/FP32"

    processor = AutoProcessor.from_pretrained(model_id, trust_remote_code=trc)
    tokenizer = None
elif model_arch == "nanollava":
    model_id = "qnguyen3/nanoLLaVA"
    model_path = models_parent_dir / "nanoLLaVA/FP32"
    trc = True

    config = AutoConfig.from_pretrained(model_id, trust_remote_code=trc)
    processor = AutoProcessor.from_pretrained(config.mm_vision_tower, trust_remote_code=trc)
    tokenizer = AutoTokenizer.from_pretrained(model_id, trsut_remote_code=trc)
elif model_arch == "minicpmv":
    model_id = "openbmb/MiniCPM-V-2_6"
    model_path = models_parent_dir / "MiniCPM-V-2_6/FP32"
    trc = True

    processor = AutoProcessor.from_pretrained(model_id, trust_remote_code=trc)
    tokenizer = None
else:
    raise Exception

quantization_config.trust_remote_code = trc

if not model_path.exists():
    OVModelForVisualCausalLM.from_pretrained(
        model_id, export=True, trust_remote_code=trc, load_in_8bit=False, compile=False
    ).save_pretrained(model_path)
    processor.save_pretrained(model_path)
    if tokenizer is not None:
        tokenizer.save_pretrained(model_path)

if compress:
    model = OVModelForVisualCausalLM.from_pretrained(
        model_path, trust_remote_code=trc, quantization_config=quantization_config
    )
    compressed_model_path = model_path / "../int4"
    model.save_pretrained(compressed_model_path)
    model = OVModelForVisualCausalLM.from_pretrained(compressed_model_path, trust_remote_code=trc)
else:
    model = OVModelForVisualCausalLM.from_pretrained(model_path, trust_remote_code=trc)

image_file = "http://images.cocodataset.org/val2017/000000039769.jpg"
raw_image = Image.open(requests.get(image_file, stream=True).raw)
inputs = model.preprocess_inputs(processor, "What are these?", raw_image, tokenizer=tokenizer)

if model_arch == "nanollava":
    streamer = TextStreamer(tokenizer, skip_prompt=True, skip_special_tokens=True)
    output_ids = model.generate(**inputs, max_new_tokens=128, use_cache=True, streamer=streamer)
else:
    output = model.generate(**inputs, max_new_tokens=10, do_sample=False)
    print(processor.decode(output[0], skip_special_tokens=True))
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/intel/openvino/modeling_visual_language.py [hunk 1]
```diff
@@ -695,6 +725,18 @@ def can_generate(self):
         """Returns True to validate the check that the model using `GenerationMixin.generate()` can indeed generate."""
         return True
 
+    @staticmethod
+    @abstractmethod
+    def assemble_input(
```
Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_visual_language.py:730]: Can we change it to `prepare_inputs()` everywhere?

Comment (AlexKoff88) [reply optimum/intel/openvino/modeling_visual_language.py:730]: `preprocess_inputs()`

**Comment by AlexKoff88** [comment]
@eaidova, @l-bat, please help with review

File: optimum/intel/openvino/modeling_visual_language.py (hunk 1 above)
Comment (nikita-savelyevv) [reply optimum/intel/openvino/modeling_visual_language.py]: Done

File: optimum/intel/openvino/quantization.py [hunk 2]
```diff
@@ -421,6 +430,14 @@ def _quantize_ovbasemodel(
                     for sub_model in sub_models:
                         _weight_only_quantization(sub_model.model, quantization_config)
                     self.model.clear_requests()
+                elif isinstance(self.model, OVModelForVisualCausalLM):
+                    language_model = self.model.language_model
+                    _weight_only_quantization(language_model.model, quantization_config, calibration_dataset)
+                    sub_model_names = ["vision_embeddings", "text_embeddings"] + self.model.additional_parts
+                    sub_models = [getattr(self.model, f"{name}_model") for name in sub_model_names]
+                    for sub_model in sub_models:
+                        _weight_only_quantization(sub_model, OVWeightQuantizationConfig(bits=8, sym=False))
```
Comment (l-bat) [inline optimum/intel/openvino/quantization.py:439]: Why do you apply INT8_ASYM data-free quantization to the sub-models?
I'm just curious about the size of these models and whether you have observed any performance improvements.

File: optimum/intel/openvino/quantization.py [hunk 3]
```diff
@@ -733,6 +750,68 @@ def _prepare_causal_lm_dataset(self, quantization_config: OVWeightQuantizationCo
...
+            )
+
+            language_model_inputs = self.model.language_model.prepare_inputs(
+                input_ids=None,
+                attention_mask=attention_mask,
+                position_ids=position_ids,
+                inputs_embeds=inputs_embeds,
+            )
+
+            pbar.update(1)
```
Comment (l-bat) [inline optimum/intel/openvino/quantization.py:807]: Why do you need to update pbar manually?
Why not just use `tqdm(dataset, ...)`?

File: optimum/commands/export/openvino.py [hunk 4]
```diff
@@ -329,11 +329,18 @@ def run(self):
             model.save_pretrained(self.args.output)
             if not self.args.disable_convert_tokenizer:
                 maybe_convert_tokenizers(library_name, self.args.output, model, task=task)
-        elif task.startswith("text-generation") and quantize_with_dataset:
-            from optimum.intel import OVModelForCausalLM
+        elif (task.startswith("text-generation") or task == "image-text-to-text") and quantize_with_dataset:
+            if task.startswith("text-generation"):
```
Comment (eaidova) [inline optimum/commands/export/openvino.py:333]: not an issue, just an idea, we may have some mapping between tesk and model inference class in future to simplify initialization process. Additionally we can provide message with recommendation which class to use orr create OVAutoModel class that will load model to specific task based on exported model config (using task and info about model type if necessary)

File: optimum/intel/openvino/modeling_visual_language.py [hunk 5]
```diff
@@ -287,6 +299,18 @@ def __init__(
         except AttributeError:
             pass
 
+    def clear_requests(self):
```
Comment (eaidova) [inline optimum/intel/openvino/modeling_visual_language.py:302]: I suggest to move compilation, nameing related features that are not directly relate to quantization into this pr https://github.com/huggingface/optimum-intel/pull/967
part of them already fixed, so I can join additional changes provide from your side too and provide tests for basic api

File: optimum/intel/openvino/modeling_visual_language.py [hunk 6]
```diff
@@ -864,6 +906,20 @@ def _filter_unattended_tokens(self, input_ids, attention_mask, past_key_values):
         position_ids = torch.sum(attention_mask, dim=1).unsqueeze(-1) - 1
         return attention_mask, position_ids
 
+    @staticmethod
+    def preprocess_inputs(
+        processor,
+        instruction: str,
```
Comment (eaidova) [inline optimum/intel/openvino/modeling_visual_language.py:912]:
```suggestion
       text: str,
```
that is not necessary can be insturction

File: optimum/intel/openvino/modeling_visual_language.py (hunk 5 above)
Comment (nikita-savelyevv) [reply optimum/intel/openvino/modeling_visual_language.py:302]: @eaidova This is fine with me. But we'll have to merge #967 before this PR.

File: optimum/intel/openvino/quantization.py (hunk 3 above)
Comment (nikita-savelyevv) [reply optimum/intel/openvino/quantization.py:807]: Thanks, fixed

File: optimum/intel/openvino/modeling_visual_language.py (hunk 5 above)
Comment (eaidova) [reply optimum/intel/openvino/modeling_visual_language.py:302]: just to not bring complexity, Let's merge your changes fist, then I'll update remained part in my PR

**Review by eaidova — APPROVED** [review]

File: optimum/intel/openvino/quantization.py (hunk 2 above)
Comment (nikita-savelyevv) [reply optimum/intel/openvino/quantization.py:439]: > Why do you apply INT8_ASYM data-free quantization to the sub-models?

This is what we've agreed on with @AlexKoff88 beforehand. 

> I'm just curious about the size of these models and whether you have observed any performance improvements.

For the VLMs in the description, text/vision embedding model sizes are in range of 150-500M parameters.

I've compared the inference speed for the two cases when we keep submodels in FP32 or compress to INT8_ASYM. Locally, I observe that the compressed scenario is about 5% slower. Maybe the situation is different on GPU devices.

But as I remember, @eaidova has mentioned, that on GPU there are some issues, when text/vision embeddings are inferred in INT8. So I'm inclined to keep them in FP32. 

Another option is to introduce some parameter threshold. For example, by default we compress only 1B+ LLMs.

@AlexKoff88 What do you think?

Comment (AlexKoff88) [reply optimum/intel/openvino/quantization.py:440]: WIll be done in the follow-up PR.

**Pull request merged**